django_modern_rest/serialization.py

```diff
@@ -124,3 +126,177 @@ def optimize_endpoint(cls, metadata: 'EndpointMetadata') -> None:
+        context = self._collect_context(controller, request, *args, **kwargs)
+        validated = self._validate_context(context)
+        self._bind_parsed(controller, validated)
+
+    @classmethod
+    def for_controller(
+        cls,
+        controller_cls: type[Any],
+        serializer: type[BaseSerializer],
+    ) -> 'SerializerContext':
+        """Return controller cached context instance (built at import-time)."""
+        return cast(
+            SerializerContext,
+            controller_cls._serializer_context,  # noqa: SLF001
+        )
+
+    @classmethod
+    def build_for_class(
+        cls,
+        controller_cls: type[Any],
+        serializer: type[BaseSerializer],
+    ) -> 'SerializerContext':
+        """Eagerly build context for a given controller and serializer."""
+        specs: list[_ComponentSpec] = []
+        names: list[str] = []
+        type_map = _build_type_map(controller_cls, serializer, specs, names)
+
+        combined_name = f'_{controller_cls.__name__}ContextModel'
+        CombinedModel = TypedDict(combined_name, type_map, total=True)  # type: ignore[misc]
+
+        return SerializerContext(
+            tuple(specs),
+            serializer,
+            CombinedModel,
+            tuple(names),
+        )
+
+    def _collect_context(
+        self,
+        controller: Any,
+        request: HttpRequest,
+        *args: Any,
+        **kwargs: Any,
+    ) -> dict[str, Any]:
+        """Collect raw data for all components into a mapping."""
+        context: dict[str, Any] = {}
+        for spec in self._specs:
+            raw = spec.data_getter(
+                controller,
+                self._serializer,
+                (spec.model,),
+                request,
+                *args,
+                **kwargs,
+            )
+            context[spec.name] = raw
+        return context
+
+    def _validate_context(self, context: dict[str, Any]) -> dict[str, Any]:
+        """Validate the combined payload using the cached TypedDict model."""
+        try:  # noqa: WPS229
+            return self._validate_via_serializer(context)
+        except self._serializer.validation_error as exc:
+            errors = self._try_strip_component_prefix(
+                self._serializer.error_to_json(exc),
+            )
+            raise RequestSerializationError(errors) from None
+        except Exception as exc:
+            raise RequestSerializationError(str(exc)) from exc
+
+    def _validate_via_serializer(
+        self,
+        context: dict[str, Any],
+    ) -> dict[str, Any]:
+        """Run serializer validation for the combined model."""
+        return cast(
+            dict[str, Any],
+            self._serializer.from_python(
+                context,
+                self._combined_model,
+                strict=False,
+            ),
+        )
+
+    def _try_strip_component_prefix(self, errors: list[Any]) -> list[Any]:
```

Comment:
We don't need to do that. Any error message we produce is good enough. This happens in runtime, so please prefer speed over nice little features.